### The goal of this study is to generate a value proposition for a new product after understanding which authors and books are on top of the 'charts'.

### Task
#### -Find the number of books released after January 1, 2000.
#### -Find the number of user reviews and the average rating for each book.
#### -Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).
#### -Identify the author with the highest average book rating: look only at books with at least 50 ratings.
#### -Find the average number of text reviews among users who rated more than 50 books.

In [1]:
import pandas as pd
from sqlalchemy import create_engine

In [2]:
db_config = {'user': 'practicum_student',
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs',
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,
             'db': 'data-analyst-final-project-db'} 

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

Unfolding the data.

In [3]:
def func_display(x):
    x = pd.io.sql.read_sql(x, con = engine)
    display(x.head(10))

In [4]:
books = '''
SELECT *
FROM books
'''
func_display(books)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

In [5]:
authors = '''
SELECT *
FROM authors
'''
func_display(authors)

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

In [6]:
publishers = '''
SELECT *
FROM publishers
'''
func_display(publishers)

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

In [7]:
ratings = '''
SELECT *
FROM ratings
'''
func_display(ratings)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

In [8]:
reviews = '''
SELECT *
FROM reviews
'''
func_display(reviews)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

In [9]:
query_0 = '''
SELECT COUNT(book_id)
FROM books
AS books_number
WHERE publication_date > '2000-01-01'
'''
books_count_after = pd.read_sql(query_0, engine)
print(books_count_after)

   count
0    819


The number of books that were published after January first 2000 (not until 2023).

In [10]:
query_1 = '''
SELECT COUNT(DISTINCT review_id)
AS reviews_count, AVG(rating) 
AS avg_rating, books.title
FROM books
INNER JOIN ratings ON ratings.book_id = books.book_id
INNER JOIN reviews ON reviews.book_id = ratings.book_id
GROUP BY books.title
ORDER BY reviews_count DESC
'''
stat = pd.read_sql(query_1, engine)
display(stat)

reviews_count  avg_rating  \
0                8    4.138462   
1                7    3.662500   
2                6    3.830508   
3                6    3.772727   
4                6    4.414634   
..             ...         ...   
988              1    3.500000   
989              1    3.000000   
990              1    4.000000   
991              1    3.500000   
992              1    4.000000   

                                                 title  
0                                  Memoirs of a Geisha  
1                              Twilight (Twilight  #1)  
2               The Da Vinci Code (Robert Langdon  #2)  
3                                             The Road  
4    Harry Potter and the Prisoner of Azkaban (Harr...  
..                                                 ...  
988                  Naked Empire (Sword of Truth  #8)  
989                                Moo  Baa  La La La!  
990                Merrick (The Vampire Chronicles #7)  
991                                          Babyville  
992                                         Lysistrata  

[993 rows x 3 columns]

'Memoirs of a Geisha' is leading with number of reviews, 'Twilight'(team Jakob for the win) is second. Reviews aren't that common with many 5 avg rating of 1 review (i would eliminate(at least 2 reviews).

In [11]:
query_2 = '''
SELECT publisher, COUNT(books.book_id)
FROM publishers
INNER JOIN books ON books.publisher_id = publishers.publisher_id
WHERE num_pages > 50
GROUP BY publishers.publisher
ORDER BY COUNT(book_id) DESC
LIMIT 1
'''
publisher_name = pd.read_sql(query_2, engine)
display(publisher_name)

publisher  count
0  Penguin Books     42

Penguin is dominating with several different publishers(maybe they're not under the same company).

In [12]:
query_3 = '''
SELECT author, avg_rating
FROM (SELECT 
author, AVG(rating) as avg_rating
FROM authors
INNER JOIN books ON books.author_id = authors.author_id
INNER JOIN ratings ON ratings.book_id = books.book_id
GROUP BY author
HAVING COUNT(ratings.book_id) > 49) as subquery
LIMIT 1
'''
top_author = pd.read_sql(query_3, engine)
display(top_author)

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.288462

J.K. Rowling A.K.A Mary GrandPré is considered (by this dataset) to have the highest average book rating('The Half Blood Price' is my favorite).

In [13]:
query_4 = '''
SELECT AVG(review_number)
FROM (SELECT 
COUNT(DISTINCT reviews.review_id) as review_number
FROM ratings
INNER JOIN reviews ON reviews.username = ratings.username
GROUP BY reviews.username
HAVING COUNT(DISTINCT ratings.rating_id) > 50) as SUBQ1
'''
top_author = pd.read_sql(query_4, engine)
display(top_author)

avg
0  24.333333

I don't know if that's a lot but it's definitely more books that i've read.

It began with 819 book from the 21st century, 'Memoirs of a Geisha' is leading with 8 reviews Twilight with 7 and the rest with 6 and below. The more popular it is with the ordinary reader the more reviews it gets (from my opinion), 'Penguin Books' is the publisher with the most books published(it's a very old publishing company, and probably know quite well which books are 'worth' publishing). J.K. Rowling A.K.A Mary GrandPré	with her series at the top of the charts is leading in avg rating. And the avg reviews is still unclear if it's good or not.
My product segestion will be a site with only an exclusive members (reviewers) and their reviews on serius books that can impact the reader and will be worthy of their time and money. both authors and reviewers will be rewarded if there are a high number of shoppers in the site and perhaps podcasts and lectures.